actors/evm/src/interpreter/precompiles.rs

```diff
@@ -12,50 +14,63 @@ use num_traits::{One, Zero};
 use substrate_bn::{pairing_batch, AffineG1, AffineG2, Fq, Fq2, Fr, Group, Gt, G1, G2};
 use uint::byteorder::{ByteOrder, LE};
 
-pub fn is_precompile(addr: &H160) -> bool {
-    !addr.is_zero() && addr <= &MAX_PRECOMPILE
-}
-
 #[derive(Debug, Eq, PartialEq)]
 pub enum PrecompileError {
     EcErr,
     IncorrectInputSize,
 }
 
-pub type PrecompileFn = fn(&[u8]) -> PrecompileResult;
+pub type PrecompileFn<RT> = fn(&RT, &[u8]) -> PrecompileResult;
 pub type PrecompileResult = Result<Vec<u8>, PrecompileError>; // TODO i dont like vec
 
-/// List of precompile smart contracts, index + 1 is the address (another option is to make an enum)
-pub const PRECOMPILES: [PrecompileFn; 9] = [
-    ec_recover, // ecrecover 0x01
-    sha256,     // SHA256 (Keccak) 0x02
-    ripemd160,  // ripemd160 0x03
-    identity,   // identity 0x04
-    modexp,     // modexp 0x05
-    ec_add,     // ecAdd 0x06
-    ec_mul,     // ecMul 0x07
-    ec_pairing, // ecPairing 0x08
-    blake2f,    // blake2f 0x09
-];
-
-pub const MAX_PRECOMPILE: H160 = {
-    let mut bytes = [0u8; 20];
-    bytes[0] = PRECOMPILES.len() as u8;
-    H160(bytes)
-};
+/// Generates a list of precompile smart contracts, index + 1 is the address (another option is to make an enum)
+const fn gen_precompiles<RT: Primitives>() -> [PrecompileFn<RT>; 9] {
+    [
+        ec_recover, // ecrecover 0x01
+        sha256,     // SHA256 (Keccak) 0x02
+        ripemd160,  // ripemd160 0x03
+        identity,   // identity 0x04
+        modexp,     // modexp 0x05
+        ec_add,     // ecAdd 0x06
+        ec_mul,     // ecMul 0x07
+        ec_pairing, // ecPairing 0x08
+        blake2f,    // blake2f 0x09
+    ]
+}
 
-pub fn call_precompile(precompile_addr: H160, input: &[u8]) -> PrecompileResult {
-    PRECOMPILES[precompile_addr.0[0] as usize - 1](input)
+pub struct Precompiles<BS, RT>(PhantomData<BS>, PhantomData<RT>);
```

Comment:
I _think_ you can drop the blockstore from this generic.

Reply:
Ah, no, no we can't. Because we're not using an associated type for some reason?

Reply:
yeah the BS wont work without being used